environments/builder/builder.go

```diff
@@ -0,0 +1,154 @@
+package builder
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/mholt/archiver"
+
+	"github.com/fission/fission"
+)
+
+const (
+	pathPrefix      = "/tmp/fission"
+	srcPkgPath      = pathPrefix + "/srcPkg"
+	deployPkgPath   = pathPrefix + "/deployPkg"
+	archiveFilename = "archive.zip"
+
+	// supported environment variables
+	envSrcPkg    = "SRC_PKG"
+	envDeployPkg = "DEPLOY_PKG"
+)
+
+type (
+	Builder struct {
+		sharedVolumePath string
+	}
+)
+
+func MakeBuilder(sharedVolumePath string) *Builder {
+	return &Builder{
+		sharedVolumePath: sharedVolumePath,
+	}
+}
+
+func (builder *Builder) Handler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "", 404)
+		return
+	}
+
+	startTime := time.Now()
+	defer func() {
+		elapsed := time.Now().Sub(startTime)
+		log.Printf("elapsed time in build request = %v", elapsed)
+	}()
+
+	// parse request
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Error reading request body")
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	var req fission.PackageBuildRequest
+	err = json.Unmarshal(body, &req)
+	if err != nil {
+		log.Printf("Error reading request body: %v", err)
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	log.Printf("builder received request: %v", req)
+
+	srcPkgArchivePath := filepath.Join(builder.sharedVolumePath, req.SrcPkgFilename)
+	err = builder.unarchive(srcPkgArchivePath, srcPkgPath)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to unarchive source package: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	buildLogs, err := builder.build(req.BuildCommand, srcPkgPath)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to build source package: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	archivePath := filepath.Join(builder.sharedVolumePath, archiveFilename)
+	err = builder.archive(deployPkgPath, archivePath)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to archive deployment package: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	resp := fission.PackageBuildResponse{
+		ArchiveFilename: archiveFilename,
+		BuildLogs:       buildLogs,
+	}
+
+	rBody, err := json.Marshal(resp)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to encode response body: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	w.Write(rBody)
+	w.WriteHeader(http.StatusOK)
+}
+
+func (builder *Builder) build(command string, workDir string) (string, error) {
+	// use `/bin/sh -c` to run multiple commands at the same time
```

Comment:
comment seems unnecessary, you're only running one command?

Reply:
User may use `&&` to concatenate multiple commands into one build command. In this case, we need to use `/bin/sh -c` to run them at once. For example:
```
pip3 install -r ${SRC_PKG}/requirements.txt -t ${SRC_PKG} && cp -r ${SRC_PKG} ${DEPLOY_PKG}
```
 But I can remove the comment if it's unnecessary.

Reply:
well this comment won't be seen by users... also it would be easier to use a script inside the container instead of making a complex build command like this.